src/expr/type_node.h

```diff
@@ -530,6 +548,38 @@ class CVC5_EXPORT TypeNode
    */
   bool isFunctionLike() const;
 
+  /**
+   * Is instance of, returns true if this type is equivalent to the
+   * leastUpperBound (see TypeNode::leastUpperBound) of itself and t.
+   */
+  bool isInstanceOf(const TypeNode& t) const;
+  /**
+   * Is comparable to type t, returns true if this type and t have a non-null
+   * leastUpperBound (see TypeNode::leastUpperBound).
+   */
+  bool isComparableTo(const TypeNode& t) const;
+  /**
+   * Least upper bound with type.
+   *
+   * We consider a partial order on types such that T1 <= T2 if T2 is an
```

Comment:
I guess the partial order defined this way to guarantee `?` is the minimal element.